File: Algorithms/850/rectangle-area-ii.cpp
class Solution {
public:
    int rectangleArea(vector<vector<int>>& rectangles) {
        long long opt = 0;
        long long mod = 1000000007;
        
        while (rectangles.size()) {
            auto area = ((rectangles[0][2] - rectangles[0][0])%mod) * ((rectangles[0][3] - rectangles[0][1])%mod);
            opt += area % mod;
            opt %= mod;
            
            for (int i = 1, ed = (int)rectangles.size(); i < ed; ++i) {
                auto subRecAry = subRec(rectangles[0], rectangles[i]);
                if (subRecAry.size() > 0) {
                    rectangles[i] = subRecAry[0];
                    for (int j = 1; j < subRecAry.size(); ++j) {
                        rectangles.push_back(subRecAry[j]);
                    }
                }
            }
            
            rectangles.erase(rectangles.begin());
        }
        
        return (int)opt;
    }
    
    vector<int> subLine(vector<int> line1, vector<int> line2) {
        vector<int> optAry = {max(line1[0], line2[0]), min(line1[1], line2[1])};
        if (optAry[0] < optAry[1]) {
            return optAry;
        }
        
        return {};
    }
    
    vector<vector<int>> subRec(vector<int> rec1, vector<int> rec2) {
        auto x = subLine({rec1[0], rec1[2]}, {rec2[0], rec2[2]});
        if (x.size() <= 0) return {};
        auto y = subLine({rec1[1], rec1[3]}, {rec2[1], rec2[3]});
        if (y.size() <= 0) return {};
        
        vector<vector<int>> optAry;
        
        vector<int> rec = {rec2[0], rec2[1], x[0], rec2[3]};// left
        if (rec[0] < rec[2] && rec[1] < rec[3]) {
            optAry.push_back(rec);
        }
        
        rec = {x[0], y[1], x[1], rec2[3]};// up
        if (rec[0] < rec[2] && rec[1] < rec[3]) {
            optAry.push_back(rec);
        }
        
        rec = {x[0], rec2[1], x[1], y[0]};// down
        if (rec[0] < rec[2] && rec[1] < rec[3]) {
            optAry.push_back(rec);
        }
        
        rec = {x[1], rec2[1], rec2[2], rec2[3]};// right
        if (rec[0] < rec[2] && rec[1] < rec[3]) {
            optAry.push_back(rec);
        }
        
        if (optAry.size() <= 0) {
            optAry.push_back({0, 0, 0, 0});
        }
        return optAry;
    }
};

File: Algorithms/[1, 1k)/850/rectangle-area-ii.swift
class Solution {
    
    func rectangleArea(_ rectangles: [[Int]]) -> Int {
        let mod: Int64 = 1_0000_0000_7
        var recAry = rectangles
        var opt: Int64 = 0
        
        while recAry.count > 0 {
            let area = Int64(recAry[0][2] - recAry[0][0]) * Int64(recAry[0][3] - recAry[0][1])
            opt += area % mod
            opt %= mod
            
            var i = 1;
            let ed = recAry.count
            while i < ed {
                let subRecAry = self.subRec(rec1: recAry[0], rec2: recAry[i])
                if subRecAry.count >= 1 {
                    recAry[i] = subRecAry.first!
                    recAry += subRecAry[1..<subRecAry.count]
                }
                i += 1
            }
            
            recAry.removeFirst()
        }
        return Int(opt);
    }
    
    func subLine(line1: [Int], line2: [Int]) -> [Int] {
        let opt = [max(line1[0], line2[0]), min(line1[1], line2[1])]
        return opt[0] < opt[1] ? opt : []
    }
    
    func subRec(rec1: [Int], rec2: [Int]) -> [[Int]] {
        let x = self.subLine(line1: [rec1[0], rec1[2]], line2: [rec2[0], rec2[2]])
        if x.count <= 0 {
            return []
        }
        let y = self.subLine(line1: [rec1[1], rec1[3]], line2: [rec2[1], rec2[3]])
        if y.count <= 0 {
            return []
        }
        
        var optAry = [[Int]]()
        
        var subRec = [rec2[0], rec2[1], x[0], rec2[3]]// left
        if subRec[0] < subRec[2] && subRec[1] < subRec[3] {
            optAry.append(subRec)
        }
        
        subRec = [x[0], y[1], x[1], rec2[3]]// up
        if subRec[0] < subRec[2] && subRec[1] < subRec[3] {
            optAry.append(subRec)
        }
        
        subRec = [x[0], rec2[1], x[1], y[0]]// down
        if subRec[0] < subRec[2] && subRec[1] < subRec[3] {
            optAry.append(subRec)
        }
        
        subRec = [x[1], rec2[1], rec2[2], rec2[3]]// right
        if subRec[0] < subRec[2] && subRec[1] < subRec[3] {
            optAry.append(subRec)
        }
        
        if optAry.count <= 0 {
            optAry.append([0, 0, 0, 0])
        }
        return optAry
    }
}